### PR TITLE
introduce the prepare-for-release.sh script

### LINKED ISSUE #3399: Add version bump script to make releases easier
/kind feature

I added a placeholder here to implement `make bump-version` that will be part of the release process:
* https://github.com/kserve/kserve/blob/master/Makefile#L193
* https://github.com/kserve/kserve/blob/master/.github/workflows/automated-release.yml#L44

### PR DESCRIPTION
**Pull request by spolti**
chore:	The purpose of this script is to facilitate the
	release process by updating the KServe version
	everywhere that is necessary.

fixes https://github.com/kserve/kserve/issues/3399

THis PR was used as example: https://github.com/kserve/kserve/pull/3970/files

Usage:
```bash 
make bump-version NEW_VERSION=0.14.0-rc2 PRIOR_VERSION=0.14.0-rc1
```

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Would you like to test this in your fork?

**Comment by spolti** [comment]
> Would you like to test this in your fork?

Sure, I have tested it step by step based on this PR https://github.com/kserve/kserve/pull/3970/files

Let me give a try.

**Comment by spolti** [comment]
updating from `0.14.0-rc1` to `0.14.0`: https://github.com/spolti/kserve/commit/25e2d119965419defbe852de9edc447fd0fe462d


updating from `0.14.0` to `0.14.1-rc1`:  https://github.com/spolti/kserve/commit/0a2627c8d1e4dadc826d65522e1e3a94c7e1c0f0

File: docs/samples/explanation/alibi/alibiexplainer/pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "alibiexplainer"
-version = "0.13.0rc0"
+version = "0.14.0rc1"
```
Comment (terrytangyuan) [inline docs/samples/explanation/alibi/alibiexplainer/pyproject.toml:3]: Is this change intentional in this PR?

Comment (spolti) [reply docs/samples/explanation/alibi/alibiexplainer/pyproject.toml:3]: yes, this was outdated.

**Review by terrytangyuan — APPROVED** [review]
Thanks!

/lgtm
/approve

**Review by terrytangyuan — CHANGES_REQUESTED** [review]
Could you also update the release doc? https://github.com/kserve/kserve/blob/master/release/RELEASE_PROCESS_v2.md

cc @yuzisun

**Comment by spolti** [comment]
something wrong with CI:
```
No VM guests are running outdated hypervisor (qemu) binaries on this host.
/usr/bin/lsb_release --short --codename
noble
Error: Unexpected HTTP response: 404
```

**Comment by spolti** [comment]
> Could you also update the release doc? https://github.com/kserve/kserve/blob/master/release/RELEASE_PROCESS_v2.md

@terrytangyuan done.

File: release/RELEASE_PROCESS_v2.md [hunk 2]
```diff
@@ -47,6 +47,9 @@ Create a branch from the master and do the following:
     5. [Chart.yaml in kserve-resources](../charts/kserve-resources/Chart.yaml#L3) to `v${MAJOR}.${MINOR}.${PATCH}-rc${RELEASE_CANDIDATE_VERSION}`
     6. [values.yaml in kserve-resources](../charts/kserve-resources/values.yaml#L2) to `v${MAJOR}.${MINOR}.${PATCH}-rc${RELEASE_CANDIDATE_VERSION}`. In addition, bump `modelmeshVersion` if it a newer version of ModelMesh is available.
     7. Run `pre-commit run` to update the Helm chart documentation 
+    8. The steps are automated in the script [prepare-for-release.sh](../hack/prepare-for-release.sh)
+       9. To use it execute: `make bump-version NEW_VERSION=0.14.0-rc2 PRIOR_VERSION=0.14.0-rc1`
+       10. Note the `-` in the version, keep it, the version will be updated accordingly and the dash removed when needed.
```
Comment (yuzisun) [inline release/RELEASE_PROCESS_v2.md:52]: So we can remove the previous manual steps with this new step?

**Comment by Jooho** [comment]
nice work @spolti

File: hack/prepare-for-release.sh [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+else
+  pversion=${PRIOR_VERSION}
+fi
+echo "Normalized versions for the charts badge: prior: $pversion - new: $nversion"
+
+# Charts
+echo -e "\033[32mUpdating charts...\033[0m"
+for readmeFile in `find charts -name README.md`; do
+  echo -e "\033[32mUpdating ${readmeFile}...\033[0m"
+  sed -i "s/\bv${PRIOR_VERSION}\b/v${NEW_VERSION}/g" ${readmeFile}
```
Comment (yuzisun) [inline hack/prepare-for-release.sh:87]: on Mac it gives this error
```
sed: 1: "charts/kserve-crd/READM ...": command c expects \ followed by text
```

Comment (yuzisun) [inline hack/prepare-for-release.sh:87]:
```suggestion
  sed -i -e "s/v${PRIOR_VERSION}/v${NEW_VERSION}/g" ${readmeFile}
```

File: hack/prepare-for-release.sh [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+     sed -i "s/\bv${PRIOR_VERSION}\b/v${NEW_VERSION}/g" ${yaml}
+  fi
+done
+
+# Update hack/generate-install.sh
+echo -e "\033[32mUpdating hack/generate-install.sh...\033[0m"
+sed -i "/\"v${PRIOR_VERSION}\"/a \    \"v${NEW_VERSION}\"" hack/generate-install.sh
+
+# Update hack/quick-install.sh
+echo -e "\033[32mUpdating hack/generate_install.sh...\033[0m"
```
Comment (yuzisun) [inline hack/prepare-for-release.sh]:
```suggestion
echo -e "\033[32mUpdating hack/quick-install.sh...\033[0m"
```

File: hack/prepare-for-release.sh [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  # do not interact over empty files
+  if [ ! -s "yaml" ]; then
+     echo -e "\033[32mUpdating ${yaml}...\033[0m"
+     sed -i "s/\bv${PRIOR_VERSION}\b/v${NEW_VERSION}/g" ${yaml}
+  fi
+done
+
+# Update hack/generate-install.sh
+echo -e "\033[32mUpdating hack/generate-install.sh...\033[0m"
+sed -i "/\"v${PRIOR_VERSION}\"/a \    \"v${NEW_VERSION}\"" hack/generate-install.sh
```
Comment (yuzisun) [inline hack/prepare-for-release.sh:103]: getting error 
```
sed: 1: "/"v0.14.0"/a \    "v0.1 ...": extra characters after \ at the end of a command
```

File: release/RELEASE_PROCESS_v2.md (hunk 2 above)
Comment (spolti) [reply release/RELEASE_PROCESS_v2.md:52]: yes, but I guess it would be good to keep them in the release docs for information.

File: hack/prepare-for-release.sh (hunk 5 above)
Comment (spolti) [reply hack/prepare-for-release.sh:103]: Are you using the GNU sed on mac?

Comment (spolti) [reply hack/prepare-for-release.sh:103]: I've stopped to use the bsd one  due these kind of errors and keep compatibility with the GNU one.

File: hack/prepare-for-release.sh (hunk 3 above)
Comment (spolti) [reply hack/prepare-for-release.sh:87]: https://github.com/kserve/kserve/pull/3993#discussion_r1813034659

Comment (spolti) [reply hack/prepare-for-release.sh:87]: I guess we still need the \b, it ensures that the v${PRIOR_VERSION} is matched only if it is a whole word, not a part of another word.

File: hack/prepare-for-release.sh [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+# Update hack/quick-install.sh
```
Comment (spolti) [inline hack/prepare-for-release.sh]:
```suggestion
# Update hack/quick_install.sh
```

File: hack/prepare-for-release.sh [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+echo -e "\033[32mUpdating hack/quick-install.sh...\033[0m"
```
Comment (spolti) [inline hack/prepare-for-release.sh]:
```suggestion
echo -e "\033[32mUpdating hack/quick_install.sh...\033[0m"
```

**Comment by spolti** [comment]
@yuzisun @terrytangyuan can you please take another look?

File: hack/prepare-for-release.sh (hunk 5 above)
Comment (yuzisun) [reply hack/prepare-for-release.sh:103]: Ah I see, let's add a note then as you need the extra step to install the GNU sed

Comment (spolti) [reply hack/prepare-for-release.sh:103]: I'll try to make it compatible with both.

**Review by terrytangyuan — DISMISSED** [review]
LGTM but it would be great if someone can verify it on Mac

**Comment by spolti** [comment]
I'll test it later.

File: hack/prepare-for-release.sh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+if [[ ${PRIOR_VERSION} == *"-"* ]]; then
+  pversion=$(echo ${PRIOR_VERSION} | sed 's/-/--/g')
+else
+  pversion=${PRIOR_VERSION}
+fi
+echo "Normalized versions for the charts badge: prior: $pversion - new: $nversion"
+
+# Charts
+echo -e "\033[32mUpdating charts...\033[0m"
+for readmeFile in `find charts -name README.md`; do
```
Comment (sivanantha321) [inline hack/prepare-for-release.sh:85]: Readme files are auto generated. Can't we just generate it instead of patching it ?

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve
